### PR TITLE
Gmsh2.2

### PR DESCRIPTION
**Pull request by JensWehner**
Implement the 2.2. meshreader.

### CODE REVIEW

File: applications/Preprocessor/lib/MeshSource.h [hunk 1]
```diff
@@ -82,6 +83,9 @@ class MeshSource {
          * and the dimension of the mesh. The ordering of the nodes should
          * follow the standard hpgem ordering.
          */
+
+        size_t id;
```
Comment (Laar) [inline applications/Preprocessor/lib/MeshSource.h:87]: Same here.

**Review by Laar — COMMENTED** [review]
A few comments.

File: applications/Preprocessor/lib/MeshSource.h [hunk 2]
```diff
@@ -65,6 +65,7 @@ class MeshSource {
          * would thus get two Nodes, one for the middle node: {{0.5}} and one
          * for the periodic outer nodes: {{0}, {1}}.
          */
+        size_t id;
```
[the remark this answers is not in the document]
Comment (JensWehner) [reply applications/Preprocessor/lib/MeshSource.h:68]: at the moment the id is implicitly given by the index of the Node in the vector, which is a) really hard to remember.

The main reason at the moment is, that I need the id to reorder things without making copies.

File: applications/Preprocessor/lib/MeshSource.h (hunk 1 above)
Comment (JensWehner) [reply applications/Preprocessor/lib/MeshSource.h:87]: See above. :)

File: applications/Preprocessor/lib/MeshSource.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (Laar) [reply applications/Preprocessor/lib/MeshSource.h:68]: Is your plan to build it into the meshsource (or its user) or only in the actual reader?

[the remark this answers is not in the document]
Comment (JensWehner) [reply applications/Preprocessor/lib/MeshSource.h:68]: no idea yet. I clean it up, when the functionality is there.

**Comment by Laar** [comment]
@JensWehner I have pushed the new mesh source to the `mesh-source2` branch. Feel free to rebase/cherry-pick the commit.

It is untested, but I think it is quite straightforward to understand. If there are any problems, please let me know.

**Comment by JensWehner** [comment]
I am happy with this now.

File: applications/Preprocessor/lib/MeshSource.h [hunk 3]
```diff
@@ -159,8 +163,17 @@ class MeshSource2 {
...
+         */
+        size_t dimension;
+        /**
+         * Index of the Element
+         */
+        size_t id;
+
         bool operator==(const Element& o) const {
-            return coordinateIds == o.coordinateIds;
+            return coordinateIds == o.coordinateIds && zoneName == o.zoneName;
```
Comment (Laar) [inline applications/Preprocessor/lib/MeshSource.h:176]: Why is this needed? To me it seems quite strange if adding/removing zone names would make elements different.

File: applications/Preprocessor/lib/gmsh.cpp [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+            nodeTag - 1;  // gmsh ids are 1 indexed hpgem is zero indexed
+        temp.coordinate = std::vector<double>{x, y, z};
+        nodes_.push_back(temp);
+    }
+}
+
+void GmshReader::readElements() {
+
+    ElementReorder reorder;
+    reorder.addElementType(1, "line:", {0, 1});
```
Comment (Laar) [inline applications/Preprocessor/lib/gmsh.cpp:271]: Would again be really nice to have a link here, so that someone could later extend it easily.

File: applications/Preprocessor/lib/MeshSource.h [hunk 5]
```diff
@@ -159,8 +163,17 @@ class MeshSource2 {
          */
         std::string zoneName;
 
+        /**
+         * Dimension of the Element
+         */
+        size_t dimension;
```
Comment (Laar) [inline applications/Preprocessor/lib/MeshSource.h:169]: I think that it may be good to clarify that this is a courtesy for implementing a reader. That the a user of the source should not assume that it is set to any (sane) value.

File: applications/Preprocessor/lib/MeshSource.h [hunk 6]
```diff
@@ -159,8 +163,17 @@ class MeshSource2 {
...
+        /**
+         * Index of the Element
+         */
+        size_t id;
```
Comment (Laar) [inline applications/Preprocessor/lib/MeshSource.h:173]: I think it may be good to clarify that this is just as courtesy for implementing a reader. That this id is not required, and should not be used in any further processing steps.

File: applications/Preprocessor/lib/gmsh.cpp [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+        double real;
+        Filehandle_ >> real;
+    }
+    size_t numintegerelements;
+    Filehandle_ >> numintegerelements;
+    logger.assert_always(numintegerelements > 2,
+                         "There is no Elementspecific data in gmsh file");
+    size_t timestep;
+    size_t field_components;
+    size_t num_enteties;
```
Comment (Laar) [inline applications/Preprocessor/lib/gmsh.cpp:359]: Typo of `num_entities`?

File: applications/Preprocessor/lib/gmsh.cpp [hunk 8]
```diff
@@ -0,0 +1,446 @@
...
+        size_t dummy;
+        Filehandle_ >> dummy;
+    }
+
+    std::unordered_map<size_t, size_t> pbc_nodes;
+
+    size_t periodic_notes;
+    Filehandle_ >> periodic_notes;
+    logger(VERBOSE, "Found % periodic nodes", periodic_notes);
+    for (size_t i = 0; i < periodic_notes; i++) {
```
Comment (Laar) [inline applications/Preprocessor/lib/gmsh.cpp:398]: Type notes -> nodes

```suggestion
    size_t periodic_nodes;
    Filehandle_ >> periodic_nodes;
    logger(VERBOSE, "Found % periodic nodes", periodic_nodes);
    for (size_t i = 0; i < periodic_nodes; i++) {
```

File: applications/Preprocessor/lib/gmsh.cpp [hunk 9]
```diff
@@ -0,0 +1,446 @@
...
+        size_t node1;
+        size_t node2;
+        Filehandle_ >> node2 >> node1;
+        // gmsh is 1 indexed hpgem is zero indexed
+        pbc_nodes[node2 - 1] = node1 - 1;
+    }
+
+    // because element ids are indices into the coordinate vector for the pbcs
+    // all we have to do is renumber the node ids.
+    for (auto& node : nodes_) {
```
Comment (Laar) [inline applications/Preprocessor/lib/gmsh.cpp:408]: Is this going to work when there is more than one node that is linked up? For example, consider a square domain with periodic boundary conditions linking the top to bottom and a second one linking left to right. Then the four corner nodes will collapse to a single node. Looking at the code I'm very not sure whether this actually renumbers them all to a single nodeid.

File: applications/Preprocessor/lib/gmsh.cpp [hunk 10]
```diff
@@ -0,0 +1,446 @@
...
+        auto found = pbc_nodes.find(node.nodeId);
+        if (found != pbc_nodes.end()) {
+            node.nodeId = found->second;
+        }
+    }
+}
+
+void GmshReader::purgeLowerDimElements() {
+
+    // gmsh elements can be of any dimension smaller than the mesh dimension
```
Comment (Laar) [inline applications/Preprocessor/lib/gmsh.cpp:418]: Hmm, now the inclusion of the dimension and id into the element starts to make sense.

Purging directly would be nicer, but probably also a bit more complicated, as you would need to keep a map elementId -> element (for the included elements).

**Review by Laar — CHANGES_REQUESTED** [review]
Looks like a good start. Quite a few smaller requests of improvements.

Additionally: it may also be good to target this to master (maybe also rebase onto `meshsource2`?) As the target branch has a low priority for me.

File: applications/Preprocessor/lib/gmsh.cpp (hunk 10 above)
Comment (JensWehner) [reply applications/Preprocessor/lib/gmsh.cpp:418]: yes, I also want to keep the steps separate.

File: applications/Preprocessor/CMakeLists.txt [hunk 11]
```diff
@@ -1,48 +1,45 @@
...
+lib/FortranUnformattedFile.cpp	
+)
+	# TODO: Lacking source/include separation
+	target_include_directories(Preprocessor-lib
+		PUBLIC
+			${CMAKE_CURRENT_SOURCE_DIR}/lib
+	)
+	target_link_libraries(Preprocessor-lib
+		PUBLIC
+            HPGEM::HPGEM)
```
[the remark this answers is not in the document]
Comment (JensWehner) [reply applications/Preprocessor/CMakeLists.txt:18]: good point

File: applications/Preprocessor/lib/ElementReorder.h [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+
+#include <vector>
+#include <map>
+
+namespace Preprocessor {
+
+class ElementReorder {
+
+   public:
+    // order is a vector of indeces of the nodes in the mesh format in the order
```
[the remark this answers is not in the document]
Comment (JensWehner) [reply applications/Preprocessor/lib/ElementReorder.h:50]: yes

File: applications/Preprocessor/lib/ElementReorder.cpp [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    auto iterator =
+        std::find_if(order_per_element_.begin(), order_per_element_.end(),
+                     [&](const Element& e) {
+                         return e.dimension_ == dimension &&
+                                e.order_.size() == indeces.size();
+                     });
+    logger.assert_always(
+        iterator != order_per_element_.end(),
+        "No reordering element with dimension % and % indeces implemented.",
+        dimension, indeces.size());
```
[the remark this answers is not in the document]
Comment (JensWehner) [reply applications/Preprocessor/lib/ElementReorder.cpp:72]: good point

File: applications/Preprocessor/lib/gmsh.cpp (hunk 7 above)
Comment (JensWehner) [reply applications/Preprocessor/lib/gmsh.cpp:359]: yes

File: applications/Preprocessor/lib/gmsh.cpp (hunk 9 above)
Comment (JensWehner) [reply applications/Preprocessor/lib/gmsh.cpp:408]: I am not a 100% sure on that but, gmsh has a concept of master nodes. I think this should work. But I will check the error condition

File: applications/Preprocessor/lib/MeshSource.h (hunk 6 above)
Comment (JensWehner) [reply applications/Preprocessor/lib/MeshSource.h:173]: Yes
